File: ViewModelSHP/app/src/main/java/com/example/viewmodelshp/MyViewModel.java
package com.example.viewmodelshp;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.SavedStateHandle;

public class MyViewModel extends AndroidViewModel {
    private SavedStateHandle handle;
    private String MUMBER_KEY = getApplication().getResources().getString(R.string.NUMBER_KEY);
    private String SHP_NAME = getApplication().getResources().getString(R.string.SHP_NAME);

    public MyViewModel(@NonNull Application application, SavedStateHandle handle) {
        super(application);
        this.handle = handle;
        if (!handle.contains(MUMBER_KEY)){
            load();
        }
    }

    private void load() {
        SharedPreferences shp = getApplication().getSharedPreferences(SHP_NAME, Context.MODE_PRIVATE);
        int x = shp.getInt(MUMBER_KEY, getApplication().getResources().getInteger(R.integer.defValue));
        handle.set(MUMBER_KEY, x);
    }

    public void save(){
        SharedPreferences shp = getApplication().getSharedPreferences(SHP_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = shp.edit();
        editor.putInt(MUMBER_KEY, getNumber().getValue());
        editor.apply();
    }

    public MutableLiveData<Integer> getNumber(){
        return handle.getLiveData(MUMBER_KEY);
    }

    public void add(int x){
        handle.set(MUMBER_KEY, getNumber().getValue() + x);
    }

}

File: AndroidViewModel/app/src/main/java/com/example/androidviewmodel/MyViewModel.kt
package com.example.androidviewmodel

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.SavedStateHandle

class MyViewModel(application: Application, private val handle: SavedStateHandle) :
    AndroidViewModel(application) {
    private val key = application.resources.getString(R.string.data_key)
    private val shpName = application.resources.getString(R.string.shp_name)

    init {
        if (!handle.contains(key)) {
            load()
        }
    }

    fun getNumber(): LiveData<Int> {
        return handle.getLiveData(key)
    }

    private fun load() {
        val shp = getApplication<Application>().getSharedPreferences(shpName, Context.MODE_PRIVATE)
        val x = shp.getInt(key, 0)
        handle.set(key, x)
    }

    fun save() {
        val shp = getApplication<Application>().getSharedPreferences(shpName, Context.MODE_PRIVATE)
        val editor = shp.edit()
        editor.putInt(key, getNumber().value!!)
        editor.apply()
    }

    fun add(x: Int) {
        handle.set(key, getNumber().value!! + x)
    }
}